File: ports/wasm-dist/js/semihosting.js
/**
 * semihosting.js — JS-side FFI for CircuitPython WASM.
 *
 * "Semihosting" is the conceptual frame: JS is the host, WASM is the
 * target.  This module provides the JS side of the shared-memory
 * communication layer.
 *
 * Two mechanisms, both via WASM linear memory (no WASI fd round-trips):
 *
 *   Event injection (JS → C):
 *     pushEvent() / pushKey() write sh_event_t records into a circular
 *     buffer in WASM linear memory.  The supervisor drains them each
 *     cp_step() via sh_drain_event_ring().
 *
 *   State reading (JS → reads C state):
 *     readState() reads the sh_state_t struct that the supervisor
 *     writes at the end of each cp_step().  No WASM export call
 *     needed — just a DataView over linear memory.
 *
 * Future bidirectional FFI (Python ↔ JS object proxies) will use
 * the jsffi/proxy pattern from the MicroPython webassembly port.
 */

/* ------------------------------------------------------------------ */
/* Constants — must match supervisor/semihosting.h                     */
/* ------------------------------------------------------------------ */

// Event types
const SH_EVT_NONE          = 0x00;
const SH_EVT_KEY_DOWN      = 0x01;
const SH_EVT_KEY_UP        = 0x02;
const SH_EVT_TIMER_FIRE    = 0x10;
const SH_EVT_FETCH_DONE    = 0x11;
const SH_EVT_HW_CHANGE     = 0x20;
const SH_EVT_PERSIST_DONE  = 0x30;
const SH_EVT_RESIZE        = 0x40;

// Sizes
const SH_STATE_SIZE  = 24;
const SH_EVENT_SIZE  = 8;

/* ------------------------------------------------------------------ */
/* Semihosting handler                                                 */
/* ------------------------------------------------------------------ */

export class Semihosting {
    constructor() {
        this.instance = null;     // Set via setInstance()
    }

    setInstance(instance) {
        this.instance = instance;
    }

    /* ---- Event injection (JS → Python) ---- */

    /**
     * Write an event directly into the linear-memory event ring.
     * No WASI fd round-trip — safe to call at any time.
     * C drains the ring in hal_step() via sh_drain_event_ring().
     */
    pushEvent(eventType, eventData, arg) {
        if (!this.instance) return;
        const exports = this.instance.exports;
        if (!exports.sh_event_ring_addr) return;

        const ringAddr = exports.sh_event_ring_addr();
        const maxEvents = exports.sh_event_ring_max();
        const mem = exports.memory.buffer;

        // Ring header: [write_idx:u32] [read_idx:u32]
        const headerView = new DataView(mem, ringAddr, 8);
        const writeIdx = headerView.getUint32(0, true);

        // Write event at (writeIdx % max) in the entries array
        const entryOffset = ringAddr + 8 + (writeIdx % maxEvents) * SH_EVENT_SIZE;
        const entryView = new DataView(mem, entryOffset, SH_EVENT_SIZE);
        entryView.setUint16(0, eventType, true);
        entryView.setUint16(2, eventData, true);
        entryView.setUint32(4, arg, true);

        // Advance write_idx
        headerView.setUint32(0, writeIdx + 1, true);
    }

    /**
     * Push a keyboard event.
     */
    pushKey(keyCode, modifiers = 0) {
        this.pushEvent(SH_EVT_KEY_DOWN, keyCode, modifiers);
    }

    /* ---- State reading (JS reads Python state) ---- */

    /**
     * Read VM state from WASM linear memory.
     * C fills the struct each cp_step(); JS reads it directly via
     * the exported sh_state_addr() pointer.  No WASI fd round-trip.
     */
    readState() {
        if (!this.instance) return null;
        const exports = this.instance.exports;
        if (!exports.sh_state_addr) return null;

        const addr = exports.sh_state_addr();
        const view = new DataView(exports.memory.buffer, addr, SH_STATE_SIZE);
        return {
            supState:    view.getUint32(0,  true),
            yieldReason: view.getUint32(4,  true),
            yieldArg:    view.getUint32(8,  true),
            frameCount:  view.getUint32(12, true),
            vmDepth:     view.getUint32(16, true),
            pendingCall: view.getUint32(20, true),
        };
    }
}

/* ---- Exports for constants (useful for host code) ---- */

export {
    SH_EVT_KEY_DOWN, SH_EVT_KEY_UP, SH_EVT_TIMER_FIRE,
    SH_EVT_FETCH_DONE, SH_EVT_HW_CHANGE, SH_EVT_PERSIST_DONE,
    SH_EVT_RESIZE,
};

File: ports/wasm/supervisor/semihosting.c
/*
 * supervisor/semihosting.c — WASM ↔ JS shared-memory FFI.
 *
 * Two mechanisms, both using WASM linear memory (no WASI fds):
 *
 *   Event ring (JS → C):
 *     JS writes sh_event_t entries directly into _event_ring via
 *     sh_event_ring_addr() and advances write_idx.  The supervisor
 *     drains them in hal_step() via sh_drain_event_ring(), calling
 *     sh_on_event() for each.  Keyboard input, timer fires, and
 *     hardware changes all arrive through this single channel.
 *
 *   State export (C → JS):
 *     The supervisor writes an sh_state_t to _export_buf at the end
 *     of each cp_step().  JS reads it via sh_state_addr() — just a
 *     DataView over linear memory, no WASM export call needed.
 *
 * Why linear memory instead of WASI fds?  WASI fd_write during
 * cp_step() corrupts memory (likely memory.buffer detach after
 * GC triggers memory.grow).  Direct memory access is safe and fast.
 */

#include "supervisor/semihosting.h"

/* ------------------------------------------------------------------ */
/* sh_init                                                             */
/* ------------------------------------------------------------------ */

void sh_init(void) {
    /* Event ring and state export use linear memory — static globals
     * are zero-initialized, nothing to open.  Kept as a lifecycle
     * hook for future FFI init (e.g., jsffi proxy tables). */
}

/* ------------------------------------------------------------------ */
/* Event dispatch — weak default                                       */
/*                                                                     */
/* supervisor.c overrides this to route events:                        */
/*   KEY_DOWN → _rx_buf, Ctrl-C → mp_sched_keyboard_interrupt, etc.   */
/* ------------------------------------------------------------------ */

__attribute__((weak))
void sh_on_event(const sh_event_t *evt) {
    (void)evt;
}

/* ------------------------------------------------------------------ */
/* Linear-memory event ring                                            */
/*                                                                     */
/* JS writes events directly into this buffer via sh_event_ring_addr() */
/* and advances write_idx.  C reads and advances read_idx.  No WASI    */
/* fd calls — safe for per-frame use in cp_step().                     */
/* ------------------------------------------------------------------ */

static struct {
    uint32_t write_idx;  /* JS increments after writing an event */
    uint32_t read_idx;   /* C increments after reading an event */
    sh_event_t entries[SH_EVENT_MAX];
} _event_ring;

__attribute__((export_name("sh_event_ring_addr")))
uintptr_t sh_event_ring_addr(void) {
    return (uintptr_t)&_event_ring;
}

__attribute__((export_name("sh_event_ring_max")))
uint32_t sh_event_ring_max(void) {
    return SH_EVENT_MAX;
}

void sh_drain_event_ring(void) {
    while (_event_ring.read_idx != _event_ring.write_idx) {
        uint32_t idx = _event_ring.read_idx % SH_EVENT_MAX;
        sh_event_t *evt = &_event_ring.entries[idx];
        if (evt->event_type != SH_EVT_NONE) {
            sh_on_event(evt);
        }
        _event_ring.read_idx++;
    }
}

/* ------------------------------------------------------------------ */
/* sh_export_state — write VM state to linear memory                   */
/*                                                                     */
/* JS reads state directly from linear memory via sh_state_addr().     */
/* No WASI fd_write — safe to call every frame.                        */
/* ------------------------------------------------------------------ */

static sh_state_t _export_buf;

__attribute__((export_name("sh_state_addr")))
uintptr_t sh_state_addr(void) {
    return (uintptr_t)&_export_buf;
}

void sh_export_state(uint32_t sup_state, uint32_t yield_reason,
                     uint32_t yield_arg, uint32_t frame_count,
                     uint32_t vm_depth) {

    _export_buf.sup_state    = sup_state;
    _export_buf.yield_reason = yield_reason;
    _export_buf.yield_arg    = yield_arg;
    _export_buf.frame_count  = frame_count;
    _export_buf.vm_depth     = vm_depth;
    _export_buf.pending_call = 0;
}
